### ISSUE TITLE
Performance improvement:  render cache.

### ISSUE THREAD

**Issue by Gormartsen** [backdrop/backdrop-issues#1413]
_I promised to improve BD performance by introducing render cache around 1 month ago._

Finally first stable release is ready. Sometimes it boost BD **twice**.
You can see code changes [here](https://github.com/Gormartsen/backdrop/compare/5f5b590e5bc529c25264b3d1430d0cd2c5627ce4...Gormartsen:1.x)

You can test it here:
http://bd1.backdrop.expert (render cache **enabled**)
http://bd2.backdrop.expert  (render cache **disabled**)

Data generated by level module. Devel data available for authorized and anonymous users at the end of the page.

Login: admin
Pass: admin123

Settings for render cache is here:
http://bd1.backdrop.expert/admin/config/development/performance

Section: Render boost

There is to cache methods:
database - store in cache_render table 
file - store in public://render_cache folder 

File cache is better for system with op cache enabled, because generated cache file is static php files that could be stored in memory by op cache.

Try both.

Temporary I wrapped cached data into: <cached id="md5"> </cached> for testing purposes.

Render cache is not sensitive to user credential and do not cache forms for now.

Testing and feedback is required here.

Also I am hoping that we can include it into 1.3.x release next January.

_With love from Canada_

**Comment by klonos** [comment]
Sounds promising. Thanx for spending time on this. Although...

Cache enabled (bd1.backdrop.expert):

```
Page execution time was 2166.13 ms. Memory used at: devel_boot()=1.11 MB, devel_shutdown()=4.98 MB, PHP peak=5.25 MB.
Render boost called: 58 checksum check: 8ms cache store: 88ms cache get: 28ms Total: 124ms
```

Cache disabled (bd2.backdrop.expert):

```
Page execution time was 2006.19 ms. Memory used at: devel_boot()=1.11 MB, devel_shutdown()=4.96 MB, PHP peak=5.25 MB.
```

[scratches head]

**Comment by klonos** [comment]
...which version of PHP by the way? Might be worth to see this under PHP7 (now with #1386 in).

**Comment by klonos** [comment]
That was anonymous. After logging in and executing a clear cache:

```
Page execution time was 668.01 ms.
Memory used at: devel_boot()=1.25 MB, devel_shutdown()=10.31 MB, PHP peak=12.5 MB.
Render boost called: 74 checksum check: 10ms cache store: 61ms cache get: 17ms Total: 88ms 
```

```
Page execution time was 969.2 ms.
Memory used at: devel_boot()=1.25 MB, devel_shutdown()=10.29 MB, PHP peak=12.5 MB.
```

**Comment by Gormartsen** [comment]
@klonos refresh page for bd1. looks like it was a first hit to generate a cache.

PHPinfo: http://bd1.backdrop.expert/php.php

Websites are on shared hosting account right now.

**Comment by Gormartsen** [comment]
And I have no clue how you get 668.01ms 
I see next:
```
Page execution time was 78.55 ms. Memory used at: devel_boot()=1.11 MB, devel_shutdown()=4.84 MB, PHP peak=5.25 MB.
Render boost called: 4 checksum check: 2ms cache store: 0ms cache get: 2ms Total: 4ms
```

**Comment by klonos** [comment]
...and page refresh while still logged in:

```
Page execution time was 125.39 ms.
Memory used at: devel_boot()=1.12 MB, devel_shutdown()=4.91 MB, PHP peak=5.25 MB.
Render boost called: 8 checksum check: 4ms cache store: 0ms cache get: 3ms Total: 7ms
```

```
Page execution time was 240.48 ms.
Memory used at: devel_boot()=1.11 MB, devel_shutdown()=5.07 MB, PHP peak=5.25 MB.
```

**Comment by Gormartsen** [comment]
Yep, that is rendered cache returned. 125ms instead of 240ms

**Comment by klonos** [comment]
> ...I have no clue how you get 668.01ms 

Could be browser and environment specific (firefox with too many addons). I will try a vanilla Chrome next.

**Comment by klonos** [comment]
...log out (still on firefox):

```
Page execution time was 74.22 ms.
Memory used at: devel_boot()=1.11 MB, devel_shutdown()=4.83 MB, PHP peak=5.25 MB.
Render boost called: 4 checksum check: 2ms cache store: 0ms cache get: 1ms Total: 4ms 
```

```
Page execution time was 140.08 ms.
Memory used at: devel_boot()=1.11 MB, devel_shutdown()=4.89 MB, PHP peak=5.5 MB.
```

**Comment by klonos** [comment]
...page refresh while logged out:

```
Page execution time was 80.07 ms.
Memory used at: devel_boot()=1.11 MB, devel_shutdown()=4.84 MB, PHP peak=5.25 MB.
Render boost called: 4 checksum check: 2ms cache store: 0ms cache get: 2ms Total: 4ms 
```

```
Page execution time was 145.03 ms.
Memory used at: devel_boot()=1.11 MB, devel_shutdown()=4.9 MB, PHP peak=5.25 MB.
```

**Comment by klonos** [comment]
...so a bit worse for anonymous I guess.

**Comment by Gormartsen** [comment]
It is not empty server. There is +- some page execution delays based on CPU and Mysql load.

Basically we are testing in real environment where many drupal sites hosted.

**Comment by klonos** [comment]
...vanilla Chrome:

```
Page execution time was 73.92 ms.
Memory used at: devel_boot()=1.11 MB, devel_shutdown()=4.83 MB, PHP peak=5.25 MB.
Render boost called: 4 checksum check: 2ms cache store: 0ms cache get: 2ms Total: 4ms
```

```
Page execution time was 141.19 ms.
Memory used at: devel_boot()=1.11 MB, devel_shutdown()=4.9 MB, PHP peak=5.25 MB.
```

**Comment by Gormartsen** [comment]
This module give more profit on long node and comment listings. Try some node with hundred comments.

**Comment by klonos** [comment]
Consistently 2x speed!!! Wow. :clap:

**Comment by klonos** [comment]
...yep. Node 29 with 83 comments as anonymous:

```
Page execution time was 127.68 ms.
Memory used at: devel_boot()=1.11 MB, devel_shutdown()=5.83 MB, PHP peak=6.75 MB.
Render boost called: 5 checksum check: 5ms cache store: 0ms cache get: 2ms Total: 7ms
```

```
Page execution time was 232.7 ms.
Memory used at: devel_boot()=1.11 MB, devel_shutdown()=5.88 MB, PHP peak=6.75 MB.
```

**Comment by Gormartsen** [comment]
Also try out file storage. It's very unique idea. 
Instead of using database to store a cache, it store cache in php files.
Opcode cache is caching this cache files in memory.... and you get memory boost without memcache)

Also file cache storage is cleaning by cron... only old and inactive caches.

There is no way to get not valid cache, because CID (md5) based on data that is going to backdrop_render... so it is always up2date.

**Comment by klonos** [comment]
> Basically we are testing in real environment where many drupal sites hosted.

Ahh, that explains the occasional spikes of >600ms

**Comment by klonos** [comment]
Could this be applied to D7/D8 as well? Asking because our Drupal brethren will surely appreciate the contribution and it would contribute to getting less spite :wink:

**Comment by Gormartsen** [comment]
@klonos I am plaining to port it to D7 and also I will let Wim Leers know about it.
I promised it [here](http://alexrayu.com/blog/drupal-739-vs-drupal-8-rc1-performance#comment-2353405787)

**Comment by klonos** [comment]
Reading through that article and following linked ones I came across [BigPipe](https://www.drupal.org/project/big_pipe). Very impressive!!!

Is your solution similar? Is it compatible? I mean could we implement your solution *over* BigPipe? That would improve actual and perceived performance!

**Comment by Gormartsen** [comment]
it is not easy to implement BigPipe. Basically it require to completely rewrite page delivery process.

So current solution will not be applicable, because it will be different rendering structure.

Also it will be not so friendly for shared hosting.

I am planning to work with Layout module.
I already see some possibilities to boost it up, but it is not so simple to do.

**Comment by Gormartsen** [comment]
@klonos have you noticed that even with BigPipe for one node cached page it was slower than node with 83 comments on bd1 ?)

**Comment by quicksketch** [comment]
Thanks @Gormartsen, I haven't tried out the demo site but I looked through the code. I'm not totally sure this is going to work consistently. The fact that this is an "opt-out" caching mechanism instead of an "opt-in" makes it look like it's likely to break functionality unexpectedly. In the code, there are a bunch of special exceptions like Forms, Entities, Submit buttons, etc. Other exceptions will most certainly need to be made, and I could see this accidentally caching data that is not cacheable.

The use of flat-files for caching is interesting and we're working on something similar in #716. I would much rather see us provide an out-of-box file-based cache and use cache_set() and cache_get() rather than specifying a custom caching solution just for render cache. Then we could use that same optimization for the page cache, where I think we'd also see large gains on most environments (e.g. those without Memcache or Redis).

It's also already possible to support render caching in Backdrop using the `#cache` property. It's rarely used in D7 development, but it's still there and functions the same as D7. So you could already use the built-in render caching in places where it's appropriate without any modification to core at all: http://omnifik.com/blog/render-arrays-and-cache-drupal-7. However I admit that #cache is an obtuse implementation and not very DX friendly, which is probably why it's not used regularly in D7.

**Comment by Gormartsen** [comment]
@quicksketch I agree that code contain excludes for Forms, Entities and Submit code.
This data is always changing (like form_token for example) and render cache will generate cache that never called again.

That's why I had to do this exceptions, to do not waste cache storage. 

There is other way - add "#dynamic" part, so we can eliminate it from md5 and use PLACEHOLDERS to replace it by data from "#dynamic".

It would work perfectly for forms for example. SO we will not generate the same form again and again and again, but use stored cache + replace placeholders.

The main idea is... If we have the same $render_array (md5 is the same), theme and render functions should return the same result.

So instead of spending CPU on generating the same HTML code, we take it from cache storage.
That's where 2x boost came from.

> I would much rather see us provide an out-of-box file-based cache and use cache_set() and cache_get() rather than specifying a custom caching solution 

https://www.drupal.org/project/filecache - this is what you are looking for, right?

> It's also already possible to support render caching in Backdrop using the #cache property. It's rarely used in D7 development, but it's still there and functions the same as D7

The reason why #cache never get used, it's because developer has to make sure that data is cachable and ... nobody think about it. It require specific knowledge and understanding of cache mechanism.

Render cache "opt-out" mechanism is developer friendly. You just code and if your code is generating repeatable data - you get boost.

I will give you an example.

Front page contain 10 nodes with taxonomy terms (let's say "test"). BD or D7 right now will call next code for each node:

https://github.com/backdrop/backdrop/blob/1.x/core/modules/taxonomy/templates/taxonomy-term.tpl.php#L47
```php
<div class="content">
    <?php print render($content); ?>
  </div>
```

and generate html every time for the same term "test" 10 times. 
Render cache is generating md5 on each render($content) request for $content array and if it is the same $content it will return cache instead of processing PHP code to generate HTML.

 

And at the end...
This code gives up to 2x time faster delivery page. That's is really something to think about.

**Comment by Gormartsen** [comment]
PS: Render cache is working for authenticated and anonymous users and work perfectly with dynamic data, like a comments on a page.

**Comment by quicksketch** [comment]
> To do not waste cache storage, I had to do this exceptions.

It sounds like there should be flags in the render array to manually opt-out then, rather than making hard-coded exceptions. e.g. `#render_cache = FALSE`.

> The main idea is... If we have the same $render_array (md5 is the same), theme and render functions should return the same result.

I think this assumption is where we will get into trouble. Identical render arrays are *not* always the same after they've been rendered. Some render arrays vary based on information that's not within the render array itself, such as the current time or the current logged in user. If you have a data formatter that is using "Time ago" (e.g. `1 week, 2 days ago`) that is a formatter that is generated at run time and isn't in the render array (which only contains the raw time).

I just don't think we can make a change that will break things that are working currently. For starters, I suppose we could run the test suite and see what unexpectedly breaks, but this won't cover the situations in contrib that are also using render arrays in different ways.

> https://www.drupal.org/project/filecache - this is what you are looking for, right?
> This one is storing data in files instead of database. Quite different concepts.

Yeah something like that, but not exactly the same implementation. In any case, the same argument you brought up at #1398 apply here. Other storage mechanisms may be preferred (e.g. memcache, redis) rather than a file-based cache. For the initial implementation, a file-based cache like you've implemented here isn't necessary but it probably gives additional gains. Because this is just a simple cache, it should be implemented as a separate, reusable cache-backend, rather than a special implementation that is only usable in one place and not swappable.

But overall, the big issue here is how much code would break if this option were used. I get the feeling it'd be along the lines of the D7 Block Cache, which generally broke so many blocks that it was unusable. If we can ensure that this render cache won't break anything, then it's a possibility, but the assumption that a render array will always render the same way is where we're likely to see a lot of problems.

**Comment by Gormartsen** [comment]
> I think this assumption is where we will get into trouble. Identical render arrays are not always the same after they've been rendered. Some render arrays vary based on information that's not within the render array itself, such as the current time or the current logged in user. If you have a data formatter that is using "Time ago" (e.g. 1 week, 2 days ago) that is a formatter that is generated at run time and isn't in the render array (which only contains the raw time).

In this case we all have valid cache for 1 day or even for one minute. Still good. 
When time change, md5 will change and new cache will created. Old one will be deleted by cron.

> I just don't think we can make a change that will break things that are working currently. 

There is no brake. Please check a code. It will not use cached data if md5 is new.

> But overall, the big issue here is how much code would break if this option were used. I get the feeling it'd be along the lines of the D7 Block Cache, which generally broke so many blocks that it was unusable

Again, It will not brake it, because it will not used old cached data, like Block cache.

>..the assumption that a render array will always render the same way is where we're likely to see a lot of problems.

There is only one case, when it can happen and it connected with bad programming habits for Drupal (or backdrop). 
For example
 - using SQL queries in theme functions
 - using external API to get extra info on render phase
 - basically any code based not on data that was provided by arguments to theme() function.

**Comment by Gormartsen** [comment]
> It sounds like there should be flags in the render array to manually opt-out then, rather than making hard-coded exceptions. e.g. #render_cache = FALSE.

Yes, we can do so. Again, I will stress it out: If data is always different, we only get into cache storage waste situation, when cached data is never used again and just take a disk space from database or files.

But we can add the option for specific cases AND add "#render_cache_dynamic" to properly cache forms with dynamic values.

I will give an example. 
```php
$form['#render_cache_dynamic'] = array(
  'form_token',
  'form_build_id',
  '#build_id',
)
```

When we have this data, on a cache generation phase we can replace this data with placeholder, like ""

So cached data will be like:
```html
    <form action=>
....
<!-- REPLACE: form_build_id -->
<!-- REPLACE: form_token -->
</form>
```

And when we delivery cached data, we replace placeholders with real data.
It will make possible to properly cache forms.

Same with other data.

> For starters, I suppose we could run the test suite and see what unexpectedly breaks

I would love to do so.

**Comment by Gormartsen** [comment]
> such as the current time or the current logged in user. 

Yep, I agree that in this case there is possible issue. But only in one case - again bad habits, using code in tpl.php file like node.tpl.php:
```php

<?php global $user; if($user->fid == 1){ print "!!!!"} ?>

```

instead of using hook_preprocess_node()

```php
function mytheme_preprocess_node($variables){
 global $user;
  if($user->uid == 1){
    $variables['admin_logged'] = TRUE
  }
}
```

**Comment by quicksketch** [comment]
One of the reasons the Block Cache in D7 never worked right was because of assets (i.e. JS and CSS) that was added at render time. It looks like this would have the same problem, because it's only caching the HTML and not the related assets. I don't think even using `#attached` would work in this approach, but we'd also need to ensure `backdrop_add_css()` and `backdrop_add_js` work when in the theme layer.

In any case, can you either set up Travis CI to run on your branches or file a PR just to get a sense of the level of test passings.

**Comment by Gormartsen** [comment]
I would do testing. Thanks.

Talking about Css and js.

I will test it as well.

I hope we are talking about hook_theme functions, not tpl.php files, right?

It would be nice if you give me an example of module that had issue in your experience with block cache, so I will be able to direct my tests properly.


Regards,
Gor Martsen
@gormartsen

**Comment by Gormartsen** [comment]
btw. Thanks for feedback and hint. I started to port it to D7 and actually D7 started to act the way how you describe - missing JS and CSS for admin menu. Did not have time to do investigation why, but now I have an idea why.

Surprisly, BD handle it well, maybe it is because backdrop_render_page never used and Layout module call backdro_render separately to each block.

**Comment by Gormartsen** [comment]
Actually test is failing by timeout:
The job exceeded the maxmimum time limit for jobs, and has been terminated.

Will try to run test local.

**Comment by quicksketch** [comment]
It's not uncommon for theme functions or templates to use `backdrop_add_css()` (or friends). Back in Drupal 5/6 days, this was actually the recommended practice so that themers could easily swap out different CSS files when they overrode the .tpl.php files. Post D7 though, themers can override individual files through the theme's .info file, so using `backdrop_add_css()` isn't so common any more.

One approach that I've thought about but never implemented is doing some kind of static list of JS/CSS/libraries and building up a buffer of those used. Similar to `obstart()` and `obflush()` used by PHP, or the `backdrop_set_message()` approach that we use ourselves. Something like this:

```
backdrop_asset_record_start();
$html = backdrop_render($renderable_array);
$assets = backdrop_asset_record_stop();
print_r($assets);
```

Would result in:
```
array(
  'js' => array(),
  'css' => array(),
  'library' => array(),
);
```

With all JS/CSS/libraries that had been added in the list. Then you could add this list of assets separately to the cache entry. Then when pulling from the render cache instead, you'd have the list of assets as well as the HTML.

**Comment by Gormartsen** [comment]
It's actually easy to do even now.

```php

function backdrop_asset_record($method = 'start'){
  static $stored
  if($method == "start"){
    $current_css = backdrop_add_css();
  }else{
    $updates_css = backdrop_add_css();
    return array_diff($updates_css, $current_css);
  }
}

backdrop_asset_record('start');
$html = backdrop_render($renderable_array);
$assets = backdrop_asset_record('stop');
print_r($assets);

```

**Comment by Gormartsen** [comment]
I will add similar code and see if there is extra data need to be stored and was not automatically added.
Will do not best coding by adding backdrop_add_css to tpl.php files + preprocessors + theme hooks and etc.

backdrop_asset_record or similar function is not good here, because backdrop_render can be nested by render() function in tpl.php file.

**Comment by quicksketch** [comment]
Our Travis tests should now be working again, though you may need to rebase or pull to get the latest `.travis.yml` that contains the fix.

**Comment by Gormartsen** [comment]
@quicksketch tnx! I was able to process it now.

Default build without render cache enabled return Failed:
https://travis-ci.org/Gormartsen/backdrop/builds/97628265
```
System: Info file test (InfoFileCoreTest) 5 passes, 2 fails, and 0 exceptions

Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Fail      Other      system.test       1730 InfoFileCoreTest->testModuleInfoFil
    render_cache does not have a version string in the .info file.
Fail      Other      system.test       1734 InfoFileCoreTest->testModuleInfoFil
    Version set for all core modules.
```

Just pushed render_cache.info with "Required" and added default config via hook_install.

Will report a build status.

**Comment by Gormartsen** [comment]
I am closing this one in favour of #2349
